### ISSUE TITLE
Oops in zfs_range_add_reader with multiple handles open for zvol

### ISSUE THREAD

**Issue by trisk** [zfsonlinux/zfs#7520]
### System information

Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | SoftNAS
Distribution Version    | 4.0
Linux Kernel                 | 4.14.34
Architecture                 | amd64
ZFS Version                  | 0.7.8
SPL Version                  | 0.7.8

### Describe the problem you're observing

When running extended I/O testing on our tiered storage driver on top of zvols, we eventually (after several hours) encounter an oops after a `blkdev_fsync` is issued for the first logical block of a zvol, after a write using the same file descriptor. The thread where this occurs is in workqueue context.

The notable thing is that we use a vfs file descriptor for metadata access (metadata is all at the beginning of the zvol), and a `struct block_device` for issuing bios for user data without additional copies.

### Describe how to reproduce the problem
We don't have an isolated testcase for this yet, but it occurs when our migration workqueue thread is running concurrently with other I/O.

### Include any warning/errors/backtraces from the system logs

```
May  8 13:35:13 ip-10-0-0-103 kernel: BUG: unable to handle kernel paging request at ffffffffa04483d6
May  8 13:35:13 ip-10-0-0-103 kernel: IP: avl_insert+0x49/0xd0 [zavl]
May  8 13:35:13 ip-10-0-0-103 kernel: PGD 200e067 P4D 200e067 PUD 200f063 PMD 201bf5067 PTE 200a57061
May  8 13:35:13 ip-10-0-0-103 kernel: Oops: 0003 [#1] SMP PTI
May  8 13:35:13 ip-10-0-0-103 kernel: Modules linked in: target_core_iblock target_core_file target_core_pscsi btier(O) fuse(O) nfsd lockd grace nfs_acl auth_rpcgss sunrpc iscsi_target_mod target_core_mod configfs bnx2fc fcoe libfcoe sg libfc scsi_transport_fc 8021q garp stp llc ena(O) e1000 zfs(PO) zcommon(PO) znvpair(PO) icp(PO) spl(O) zavl(PO) zunicode(PO) pcspkr cirrus ttm i2c_piix4 tcp_htcp ext4 jbd2 mbcache xen_blkfront pata_acpi ata_generic crc32c_intel be2iscsi bnx2i cnic uio cxgb4i iw_cxgb4 cxgb4 ptp pps_core cxgb3i libcxgbi iw_cxgb3 ib_core ata_piix cxgb3 mdio libcxgb ipv6 crc_ccitt libiscsi_tcp qla4xxx iscsi_boot_sysfs libiscsi scsi_transport_iscsi ixgbevf dm_mirror dm_region_hash dm_log dm_mod dax [last unloaded: target_core_iblock]
May  8 13:35:13 ip-10-0-0-103 kernel: CPU: 0 PID: 13288 Comm: kworker/u30:3 Tainted: P           O    4.14.34 #1
May  8 13:35:13 ip-10-0-0-103 kernel: Hardware name: Xen HVM domU, BIOS 4.2.amazon 08/24/2006
May  8 13:35:13 ip-10-0-0-103 kernel: Workqueue: sdtiera-migrate block_migration_worker [btier]
May  8 13:35:13 ip-10-0-0-103 kernel: task: ffff88014fe58700 task.stack: ffffc9000185c000
May  8 13:35:13 ip-10-0-0-103 kernel: RIP: 0010:avl_insert+0x49/0xd0 [zavl]
May  8 13:35:13 ip-10-0-0-103 kernel: RSP: 0018:ffffc9000185f680 EFLAGS: 00010282
May  8 13:35:13 ip-10-0-0-103 kernel: RAX: ffffffffa04483d6 RBX: ffff8801d3e64d60 RCX: ffffffffa04483d7
May  8 13:35:13 ip-10-0-0-103 kernel: RDX: 0000000000000000 RSI: ffff880130c99c08 RDI: ffff8801d3e64d60
May  8 13:35:13 ip-10-0-0-103 kernel: RBP: ffffc9000185f688 R08: ffffffffa04483d6 R09: ffff8801a5bba6c0
May  8 13:35:13 ip-10-0-0-103 kernel: R10: 0000000000000000 R11: 0000000000000069 R12: ffff880130c99c00
May  8 13:35:13 ip-10-0-0-103 kernel: R13: 0000000000001000 R14: 0000000000000000 R15: 0000000000000000
May  8 13:35:13 ip-10-0-0-103 kernel: FS:  0000000000000000(0000) GS:ffff880207200000(0000) knlGS:0000000000000000
May  8 13:35:13 ip-10-0-0-103 kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
May  8 13:35:13 ip-10-0-0-103 kernel: CR2: ffffffffa04483d6 CR3: 000000000200a006 CR4: 00000000001606f0
May  8 13:35:13 ip-10-0-0-103 kernel: DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
May  8 13:35:13 ip-10-0-0-103 kernel: DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
May  8 13:35:13 ip-10-0-0-103 kernel: Call Trace:
May  8 13:35:13 ip-10-0-0-103 kernel: zfs_range_add_reader+0x16e/0x1f0 [zfs]
May  8 13:35:13 ip-10-0-0-103 kernel: ? spl_kmem_alloc+0x96/0x1a0 [spl]
May  8 13:35:13 ip-10-0-0-103 kernel: zfs_range_lock_reader+0x119/0x1c0 [zfs]
May  8 13:35:13 ip-10-0-0-103 kernel: ? spl_kmem_alloc+0x96/0x1a0 [spl]
May  8 13:35:13 ip-10-0-0-103 kernel: zfs_range_lock+0x11b/0x130 [zfs]
May  8 13:35:13 ip-10-0-0-103 kernel: zvol_get_data+0x79/0x160 [zfs]
May  8 13:35:13 ip-10-0-0-103 kernel: zil_lwb_commit+0x1dd/0x2f0 [zfs]
May  8 13:35:13 ip-10-0-0-103 kernel: zil_commit_writer+0xfc/0x290 [zfs]
May  8 13:35:13 ip-10-0-0-103 kernel: ? generic_make_request_checks+0x25a/0x550
May  8 13:35:13 ip-10-0-0-103 kernel: zil_commit+0xe2/0x150 [zfs]
May  8 13:35:13 ip-10-0-0-103 kernel: zvol_request+0x2e0/0x3a0 [zfs]
May  8 13:35:13 ip-10-0-0-103 kernel: ? __pagevec_lru_add+0x17/0x20
May  8 13:35:13 ip-10-0-0-103 kernel: generic_make_request+0x100/0x2c0
May  8 13:35:13 ip-10-0-0-103 kernel: submit_bio+0x6d/0x130
May  8 13:35:13 ip-10-0-0-103 kernel: submit_bio_wait+0x59/0x70
May  8 13:35:13 ip-10-0-0-103 kernel: blkdev_issue_flush+0x9e/0xc0
May  8 13:35:13 ip-10-0-0-103 kernel: blkdev_fsync+0x3f/0x60
May  8 13:35:13 ip-10-0-0-103 kernel: vfs_fsync_range+0x4e/0xe0
May  8 13:35:13 ip-10-0-0-103 kernel: write_device_magic+0x113/0x1e0 [btier]
May  8 13:35:13 ip-10-0-0-103 kernel: copyblock+0x146/0x1b0 [btier]
May  8 13:35:13 ip-10-0-0-103 kernel: migrate_block+0x61/0x100 [btier]
May  8 13:35:13 ip-10-0-0-103 kernel: migrate_if_needed+0xf1/0x1b0 [btier]
May  8 13:35:13 ip-10-0-0-103 kernel: block_migration_worker+0x54/0x60 [btier]
May  8 13:35:13 ip-10-0-0-103 kernel: process_one_work+0x19b/0x470
May  8 13:35:13 ip-10-0-0-103 kernel: ? __schedule+0x2ec/0x750
May  8 13:35:13 ip-10-0-0-103 kernel: ? schedule+0x3a/0xa0
May  8 13:35:13 ip-10-0-0-103 kernel: worker_thread+0x6b/0x480
May  8 13:35:13 ip-10-0-0-103 kernel: ? default_wake_function+0x12/0x20
May  8 13:35:13 ip-10-0-0-103 kernel: ? __wake_up_common+0x8a/0x130
May  8 13:35:13 ip-10-0-0-103 kernel: ? maybe_create_worker+0x140/0x140
May  8 13:35:13 ip-10-0-0-103 kernel: ? schedule+0x3a/0xa0
May  8 13:35:13 ip-10-0-0-103 kernel: ? maybe_create_worker+0x140/0x140
May  8 13:35:13 ip-10-0-0-103 kernel: kthread+0x113/0x150
May  8 13:35:13 ip-10-0-0-103 kernel: ? __kthread_init_worker+0x40/0x40
May  8 13:35:13 ip-10-0-0-103 kernel: ? __kthread_init_worker+0x40/0x40
May  8 13:35:13 ip-10-0-0-103 kernel: ret_from_fork+0x35/0x40
May  8 13:35:13 ip-10-0-0-103 kernel: Code: 00 00 48 83 e0 fe 48 63 c9 48 c7 06 00 00 00 00 48 83 c9 01 48 c7 46 08 00 00 00 00 48 09 c1 48 85 c0 48 89 4e 10 74 75 48 63 d2 <48> 89 34 d0 48 8b 70 10 89 f1 83 e1 03 83 e9 01 89 cb 03 1c 95
May  8 13:35:13 ip-10-0-0-103 kernel: RIP: avl_insert+0x49/0xd0 [zavl] RSP: ffffc9000185f680
May  8 13:35:13 ip-10-0-0-103 kernel: CR2: ffffffffa04483d6
May  8 13:35:13 ip-10-0-0-103 kernel: ---[ end trace 28dd41411409c1d4 ]---
```

**Comment by ryao** [comment]
What does this output on this system:

`echo list *avl_insert+0x49/0xd0 | gdb $(modinfo -n zavl)`

**Comment by trisk** [comment]
Should be this:
```
0x450 is in avl_insert (/usr/src/debug/zfs-kmod-0.7.8/zfs-0.7.8/module/avl/avl.c:482).
477      * After the node is inserted, a single rotation further up the tree may
478      * be necessary to maintain an acceptable AVL balance.
479      */
480     void
481     avl_insert(avl_tree_t *tree, void *new_data, avl_index_t where)
482     {
483             avl_node_t *node;
484             avl_node_t *parent = AVL_INDEX2NODE(where);
485             int old_balance;
486             int new_balance;
```

**Comment by ryao** [comment]
Unfortunately, that doesn't tell us much. Try this:

```
echo -e "display/i *avl_insert+0x49/0xd0\ndisassemble /m avl_insert+0x49/0xd0" | gdb `modinfo -n zavl`
```

**Comment by trisk** [comment]
```
Dump of assembler code for function avl_insert:                                       [2/964]
482     {
   0x0000000000000450 <+0>:     push   %rbp
   0x0000000000000451 <+1>:     mov    %rsp,%rbp
   0x0000000000000454 <+4>:     push   %rbx
   0x0000000000000455 <+5>:     callq  0x45a <avl_insert+10>

483             avl_node_t *node;
484             avl_node_t *parent = AVL_INDEX2NODE(where);
   0x0000000000000463 <+19>:    mov    %rdx,%rax
   0x0000000000000470 <+32>:    and    $0xfffffffffffffffe,%rax

485             int old_balance;
486             int new_balance;
487             int which_child = AVL_INDEX2CHILD(where);
   0x0000000000000466 <+22>:    and    $0x1,%edx

488             size_t off = tree->avl_offset;
489
490             ASSERT(tree);
491     #ifdef _LP64
492             ASSERT(((uintptr_t)new_data & 0x7) == 0);
493     #endif
494
495             node = AVL_DATA2NODE(new_data, off);
```

**Comment by ryao** [comment]
Unfortunately, the gdb output and instruction pointer do not match up. It seems that the kernel is calling `__sprint_symbol(buffer, addr, -1, 1)` while printing these addresses, such that the `symbol+offset/length` becomes `symbol+(offset-1)/(length+1)`. The comment for `__sprint_symbol` says that it is to ensure that the original caller is listed when a tail call marked noreturn is inlined, but it makes interpreting backtraces a pain if you don't know about the off by 1 possibility. Anyway, the NULL pointer dereference is this:

```
503		node->avl_child[1] = NULL;
   0x0000000000000482 <+50>:	movq   $0x0,0x8(%rsi)
```

**Comment by chrisrd** [comment]
@ryao It looks to me like the trapping instruction is:
```
510                     parent->avl_child[which_child] = node;
   0x0000000000000496 <+70>:    movslq %edx,%rdx
   0x0000000000000499 <+73>:    mov    %rsi,(%rax,%rdx,8)      <-- trapping instruction
```
...per pumping the "Code" line of the dump into `linux/scripts/decodecode`:
```
$ cat <<END | scripts/decodecode
May 8 13:35:13 ip-10-0-0-103 kernel: Code: 00 00 48 83 e0 fe 48 63 c9 48 c7 06 00 00 00 00 48 83 c9 01 48 c7 46 08 00 00 00 00 48 09 c1 48 85 c0 48 89 4e 10 74 75 48 63 d2 <48> 89 34 d0 48 8b 70 10 89 f1 83 e1 03 83 e9 01 89 cb 03 1c 95
END
May 8 13:35:13 ip-10-0-0-103 kernel: Code: 00 00 48 83 e0 fe 48 63 c9 48 c7 06 00 00 00 00 48 83 c9 01 48 c7 46 08 00 00 00 00 48 09 c1 48 85 c0 48 89 4e 10 74 75 48 63 d2 <48> 89 34 d0 48 8b 70 10 89 f1 83 e1 03 83 e9 01 89 cb 03 1c 95
All code
========
   0:   00 00                   add    %al,(%rax)
   2:   48 83 e0 fe             and    $0xfffffffffffffffe,%rax
   6:   48 63 c9                movslq %ecx,%rcx
   9:   48 c7 06 00 00 00 00    movq   $0x0,(%rsi)
  10:   48 83 c9 01             or     $0x1,%rcx
  14:   48 c7 46 08 00 00 00    movq   $0x0,0x8(%rsi)
  1b:   00
  1c:   48 09 c1                or     %rax,%rcx
  1f:   48 85 c0                test   %rax,%rax
  22:   48 89 4e 10             mov    %rcx,0x10(%rsi)
  26:   74 75                   je     0x9d
  28:   48 63 d2                movslq %edx,%rdx
  2b:*  48 89 34 d0             mov    %rsi,(%rax,%rdx,8)               <-- trapping instruction
  2f:   48 8b 70 10             mov    0x10(%rax),%rsi
  33:   89 f1                   mov    %esi,%ecx
  35:   83 e1 03                and    $0x3,%ecx
  38:   83 e9 01                sub    $0x1,%ecx
  3b:   89 cb                   mov    %ecx,%ebx
  3d:   03                      .byte 0x3
  3e:   1c 95                   sbb    $0x95,%al

Code starting with the faulting instruction
===========================================
   0:   48 89 34 d0             mov    %rsi,(%rax,%rdx,8)
   4:   48 8b 70 10             mov    0x10(%rax),%rsi
   8:   89 f1                   mov    %esi,%ecx
   a:   83 e1 03                and    $0x3,%ecx
   d:   83 e9 01                sub    $0x1,%ecx
  10:   89 cb                   mov    %ecx,%ebx
  12:   03                      .byte 0x3
  13:   1c 95                   sbb    $0x95,%al
```

That "trapping instruction" translates to "move the contents of register `%rsi` into the address pointed to by `(%rax + (%rdx * 8))`". And the register dump shows us `%rax` and `%rdx`:
```
May  8 13:35:13 ip-10-0-0-103 kernel: RAX: ffffffffa04483d6 ...
May  8 13:35:13 ip-10-0-0-103 kernel: RDX: 0000000000000000 ...
```

We can also look back at the warning and see that address in `%rax` is actually the faulting address, which all makes sense:
```
May  8 13:35:13 ip-10-0-0-103 kernel: BUG: unable to handle kernel paging request at ffffffffa04483d6
```

We can also see from the gdb dump, `%rax` is `parent`, e.g. from just before the faulting instruction:
```
508             if (parent != NULL) {
   0x000000000000048d <+61>:    test   %rax,%rax
   0x0000000000000494 <+68>:    je     0x50b <avl_insert+187>
```

And further back we see `parent` comes from `where`:
```
484             avl_node_t *parent = AVL_INDEX2NODE(where);
   0x0000000000000463 <+19>:    mov    %rdx,%rax
   0x0000000000000470 <+32>:    and    $0xfffffffffffffffe,%rax
```

Inspecting the code, we see `where` is passed in as an argument, i.e. `avl_insert()` was called with a bad `where` param. Further inspection indicates `avl_insert()` was called from:
```
static void
zfs_range_add_reader(avl_tree_t *tree, rl_t *new, rl_t *prev, avl_index_t where)
{
...
        if (next == NULL || off + len <= next->r_off) {
                /* no overlaps, use the original new rl_t in the tree */
                avl_insert(tree, new, where);
                return;
        }
...
}
```

...and `where` was passed directly into that function as well, from:

```
static void
zfs_range_lock_reader(zfs_rlock_t *zrl, rl_t *new)
{
        avl_tree_t *tree = &zrl->zr_avl;
        rl_t *prev, *next;
        avl_index_t where;
        uint64_t off = new->r_off;
        uint64_t len = new->r_len;

        /*
         * Look for any writer locks in the range.
         */
retry:
        prev = avl_find(tree, new, &where);
...
        zfs_range_add_reader(tree, new, prev, where);
}

```

After a quick inspection of `avl_find` I'm guessing the underlying problem is probably a corrupt avl tree.

As to how we got a corrupt avl tree...

**Comment by ryao** [comment]
@chrisrd I believe that I found the bug:

https://github.com/tsutsu/btier/blob/master/kernel/btier/btier_sysfs.c#L409

That should be:

```
	cur = kzalloc(p - a + 1, GFP_KERNEL);
        if (!cur)
                goto end_error;
	memcpy(cur, a, p - a);
```

That would match what is done here:

https://github.com/tsutsu/btier/blob/master/kernel/btier/btier_sysfs.c#L377

Interestingly, there are actually two bugs there. One is the memcpy is overflowing the heap object by 1 byte. The second is that error handling is missing for the case when the memory allocation fails.

I would close this, but I am going to leave it open until I have confirmation that the bug was in BTIER and not ZFS.

**Comment by chrisrd** [comment]
@ryao Nice find, hopefully that's the root cause. ...I assume you're working on the `btier` thingy with @trisk, which is how he's seeing this error from external code?

Unrelated to ZoL...

That pattern of "find a space in the source string, get a temp buffer filled with nulls, copy chars from source up to the space into the buffer, scan the buffer for an ascii number, release buffer" looks a bit odd. Why not just "scan (from the current position in the) source for an ascii number"?

**Comment by ryao** [comment]
@chrisrd I am working with @trisk on this. I also thought that the copies were unnecessary, but I decided to keep the fix simple. If I started fixing the sloppy programming, I would end up rewriting the entire codebase. The BTIER codebase is so full of issues that it is easy to spot bugs by casual inspection. If Linus Torvalds saw code like this, he would start cursing.

`get_blockinfo()` calls are not properly protected by a lock outside of btier_request.c while user discards will literally zero the `blockinfo` structures. This causes a race between `get_blockinfo()` uses and discard where the right sequence of events will cause undefined behavior. Since @trisk ported the TRIM patches from illumos to ZoL for use with BTIER at SoftNAS, he was hitting issues related to that race on top of the issues caused by the heap corruption caused by the `memcpy()`.

Also, it seems multiple other memory allocations lack error handling. Specifically, those in `wipe_bitlist()`, `tier_ioctl()` (the first one) and `as_strarrcat()`. It does, however, use error handling to check if a single `mempool_alloc()` allocation failed while ignoring 4 others. This would be bad, but all 5 use GFP_NOIO and will never return, so the error handling on the one invocation is dead code.

The driver keeps metadata and locking structures in memory for every logical block, which is incredibly wasteful and affects its ability to scale to large amounts of storage. The locking order and what various locks protect is not documented. It does not even do something so simple as group the locks and the things it protects together inside the struct definitions with empty lines to delineate what they are protecting. Its disk format uses a primitive bit array where every bit is represented by an entire byte.

The code lacks comments. There is no disk format design document (unlike ZFS, which at least has the old one that Sun published). Honestly, BTIER is the most poorly written kernel module that I have ever seen. :/
